File: 1. Sorting/python/insertion_sort.py
from typing import List

def swap(arr: List[int], i: int, j: int):
    arr[i], arr[j] = arr[j], arr[i]

def insertion_sort(arr: List[int]) -> List[int]:
    for i in range(len(arr) - 1):
        next_elem = i + 1
        while next_elem > 0 and arr[next_elem] < arr[next_elem - 1]:
            swap(arr, next_elem, next_elem - 1)
            next_elem -= 1

    return arr

if __name__ == '__main__':
    test1 = []
    test2 = [1]
    test3 = [1000, 937, 672, 321, 1, -38]
    test4 = [-838, 19999, 3, 2, 8573, -4]
    test5 = [1, 2, 3, 4, 5]

    assert(insertion_sort(test1) == [])
    assert(insertion_sort(test2) == [1])
    assert(insertion_sort(test3) == [-38, 1, 321, 672, 937, 1000])
    assert(insertion_sort(test4) == [-838, -4, 2, 3, 8573, 19999])
    assert(insertion_sort(test5) == [1, 2, 3, 4, 5])

File: 1. Sorting/cpp/insertionSort.cpp
#include <cassert>
#include <vector>
#include <iostream>

void insertionSort(std::vector<int> &arr);
void assertSorted(const std::vector<int> &, const std::vector<int> &);

int main(){
    std::vector<int> test1 = {};
    std::vector<int> test2 = {1};
    std::vector<int> test3 = {1000, 937, 672, 321, 1, -38};
    std::vector<int> test4 = {-838, 19999, 3, 2, 8573, -4};
    std::vector<int> test5 = {1, 2, 3, 4, 5};

    insertionSort(test1);
    insertionSort(test2);
    insertionSort(test3);
    insertionSort(test4);
    insertionSort(test5);

    assert( test1 == std::vector<int>{} );
    assertSorted(test2, std::vector<int> {1});
    assertSorted(test3, std::vector<int> {-38, 1, 321, 672, 937, 1000});
    assertSorted(test4, std::vector<int> {-838, -4, 2, 3, 8573, 19999});
    assertSorted(test5, std::vector<int> {1, 2, 3, 4, 5});

    return 0;
}

void insertionSort(std::vector<int> &arr){
    for(int i = 0; i < int(arr.size()); ++i){
        int curr = i;
        while(curr > 0){
            if(arr[curr] < arr[curr - 1]){
                std::swap(arr[curr], arr[curr - 1]);
            } else {
                break;
            }
            curr -= 1;
        }
    }
}

void assertSorted(const std::vector<int> &arr1, const std::vector<int> &arr2){
    assert( arr1.size() == arr2.size() );
    for(int i = 0; i < arr1.size(); ++i){
        assert( arr1[i] == arr2[i] );
    }
}